tests/utils/test_compute_node.py

```diff
@@ -0,0 +1,398 @@
+    # NOTE: Assuming that it takes more time for the job to be allocated than it takes for
+    # the job to show up in `squeue`.
+    salloc_task = asyncio.create_task(
+        sbatch(login_node_v2, sbatch_flags=allocation_flags, job_name=job_name)
+        if persist
+        else salloc(login_node_v2, salloc_flags=allocation_flags, job_name=job_name),
+        name="sbatch" if persist else "salloc",
+    )
+    get_new_job_ids_task = asyncio.create_task(
+        get_new_job_ids(), name="get_new_job_ids"
+    )
+
+    new_job_ids = await asyncio.wait_for(get_new_job_ids_task, timeout=None)
+    assert not salloc_task.done()  # hopefully we get the job ID from SQUEUE before the
+    # job is actually running...
+    salloc_task.cancel(
+        msg="Interrupting the job allocation as soon as the job ID shows up in squeue."
+    )
+    assert new_job_ids and len(new_job_ids) == 1
+    new_job_id = new_job_ids.pop()
+    # wait long enough for `squeue` to update and not show the job anymore.
+    await asyncio.sleep(10)
+    jobs_after = await get_jobs_in_squeue()
+    assert new_job_id not in jobs_after
+    assert jobs_after <= _jobs_before
+
+
+@launches_jobs
+@pytest.mark.slow
+class TestComputeNode(RunnerTests):
+    @pytest_asyncio.fixture(scope="class")
+    async def runner(
+        self, login_node_v2: RemoteV2, persist: bool, allocation_flags: list[str]
+    ):
+        if login_node_v2.hostname == "localhost":
+            pytest.skip(reason="Test doesn't currently work on the mock slurm cluster.")
+
+        if persist:
+            runner = await sbatch(
+                login_node_v2, sbatch_flags=allocation_flags, job_name="mila-code"
+            )
+        else:
+            runner = await salloc(
+                login_node_v2, salloc_flags=allocation_flags, job_name="mila-code"
+            )
+        yield runner
+        await runner.close_async()
+
+    @pytest.fixture(
+        scope="class",
+        params=[
+            ("echo OK", "OK", ""),
+            ("echo $SLURM_JOB_ID", re.compile(r"^[0-9]+"), ""),
+            ("echo $SLURM_PROCID", "0", ""),
+        ],
+    )
+    def command_and_expected_result(self, request: pytest.FixtureRequest):
+        return request.param
+
+    @pytest.mark.parametrize("use_async", [False, True], ids=["sync", "async"])
+    @pytest.mark.asyncio
+    async def test_run_gets_executed_in_job_step(
+        self, runner: ComputeNode, use_async: bool
+    ):
+        command = "echo $SLURM_STEP_ID"
+        output_a = (
+            await runner.get_output_async(command)
+            if use_async
+            else runner.get_output(command)
+        )
+        output_b = (
+            await runner.get_output_async(command)
+            if use_async
+            else runner.get_output(command)
+        )
+        job_step_a = int(output_a)
+        job_step_b = int(output_b)
+        assert job_step_a >= 0
+        assert job_step_b == job_step_a + 1
```

Comment:
Tthe goal of the test here is to check that each command is being run within a different job step - that each command is being run in a different `srun` call.
You're right that we could do a similar check using mocks, but in my opinion,  it would probably turn out to be much uglier. This also is reusing the existing test job, so It doesn't really "cost" us anything.